File: matchu_app/functions/src/recommendation/topicTaxonomy.js
"use strict";

const TOPIC_ALIASES = Object.freeze({
  "flutter-dev": "flutter",
  flutterdev: "flutter",
  dart: "flutter",
  "firebase-dev": "firebase",
  "trí-tuệ-nhân-tạo": "ai",
  "tri-tue-nhan-tao": "ai",
  "artificial-intelligence": "ai",
  "lập-trình-flutter": "flutter",
  "lap-trinh-flutter": "flutter",
});

function normalizeTopicId(value) {
  if (typeof value !== "string") return "";
  const normalized = value
    .trim()
    .toLocaleLowerCase("vi")
    .replace(/^#+/, "")
    .replace(/[_\s]+/g, "-")
    .replace(/-+/g, "-");
  return TOPIC_ALIASES[normalized] || normalized;
}

function normalizeTopicIds(values) {
  if (!Array.isArray(values)) return [];
  return [...new Set(values.map(normalizeTopicId).filter(Boolean))];
}

module.exports = { normalizeTopicId, normalizeTopicIds };

File: matchu_app/lib/utils/topic_taxonomy.dart
class TopicTaxonomy {
  TopicTaxonomy._();

  static const Map<String, String> _aliases = <String, String>{
    'flutter-dev': 'flutter',
    'flutterdev': 'flutter',
    'dart': 'flutter',
    'firebase-dev': 'firebase',
    'trí-tuệ-nhân-tạo': 'ai',
    'tri-tue-nhan-tao': 'ai',
    'artificial-intelligence': 'ai',
    'lập-trình-flutter': 'flutter',
    'lap-trinh-flutter': 'flutter',
    'lập trình flutter': 'flutter',
  };

  /// Returns stable IDs used by storage, embedding and analytics.
  static List<String> normalizeAll(Iterable<String> values) {
    final result = <String>{};
    for (final value in values) {
      final normalized = _normalize(value);
      if (normalized.isNotEmpty) result.add(_aliases[normalized] ?? normalized);
    }
    return result.toList(growable: false);
  }

  static String _normalize(String value) {
    return value
        .trim()
        .toLowerCase()
        .replaceAll(RegExp(r'^#+'), '')
        .replaceAll(RegExp(r'[_\s]+'), '-')
        .replaceAll(RegExp(r'-+'), '-');
  }
}
